modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -80,32 +76,34 @@
 
         List<InetSocketAddress> addrs = parseAddresses(clientCfg.getAddresses());
 
+        srvCnt = addrs.size();
+
         primary = addrs.get(new Random().nextInt(addrs.size())); // we already verified there is at least one address
 
         for (InetSocketAddress a : addrs) {
             if (a != primary)
-                this.backups.add(a);
+                backups.add(a);
         }
 
         ClientConnectionException lastEx = null;
 
         for (int i = 0; i < addrs.size(); i++) {
             try {
-                ch = chFactory.apply(new ClientChannelConfiguration(clientCfg).setAddress(primary)).get();
+                ch = chFactory.apply(new ClientChannelConfiguration(clientCfg).setAddress(addrs.get(i))).get();
```

Comment:
In my opinion eager connection makes sense. Problems will be detected at an earlier stage. For example, in case of some wrong server address configuration with eager connection exception will be thrown at some kind of initialization code, but with lazy initialization, we will get an exception only when business logic is running.

Reply:
Yes, such reasoning definitely makes sense. Lazy initialization can also have benefits, e.g. reducing time taken by _initialization code_. Well, seems like a trade off. I trust your own decision.

Reply:
Does Java thin client throw an exception if it can not establish connection to one of nodes?

Reply:
It throws an exception if it can't establish a connection to all nodes (exception is not thrown if connected to at least one node from the address list).
Actually, I've reverted fix for this line, since `primary` is changed in `changeServer` method and there was no error in the original implementation (I misunderstood this behavior).